specs/core/0_beacon-chain.md

```diff
@@ -1705,30 +1695,30 @@ def process_attestation(state: BeaconState, attestation: Attestation) -> None:
     attestation_slot = get_attestation_slot(state, attestation)
     assert attestation_slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot <= attestation_slot + SLOTS_PER_EPOCH
 
-    # Check target epoch, source epoch, source root, and source crosslink
     data = attestation.data
-    assert (data.target_epoch, data.source_epoch, data.source_root, data.previous_crosslink_root) in {
-        (get_current_epoch(state), state.current_justified_epoch, state.current_justified_root, hash_tree_root(state.current_crosslinks[data.shard])),
-        (get_previous_epoch(state), state.previous_justified_epoch, state.previous_justified_root, hash_tree_root(state.previous_crosslinks[data.shard])),
-    }
-
-    # Check crosslink data root
-    assert data.crosslink_data_root == ZERO_HASH  # [to be removed in phase 1]
-
-    # Check signature and bitfields
-    assert verify_indexed_attestation(state, convert_to_indexed(state, attestation))
-
-    # Cache pending attestation
     pending_attestation = PendingAttestation(
         data=data,
         aggregation_bitfield=attestation.aggregation_bitfield,
         inclusion_delay=state.slot - attestation_slot,
         proposer_index=get_beacon_proposer_index(state),
     )
+
+    assert data.target_epoch in (get_previous_epoch(state), get_current_epoch(state))
+    if data.target_epoch == get_previous_epoch(state):
+        ffg_data = (state.previous_justified_epoch, state.previous_justified_root, get_previous_epoch(state))
+        previous_crosslink = state.previous_crosslinks[data.crosslink.shard]
+        state.previous_epoch_attestations.append(pending_attestation)
     if data.target_epoch == get_current_epoch(state):
```

Comment:
`elif` would be more accurate.
```suggestion
    elif data.target_epoch == get_current_epoch(state):
```

Reply:
The logic is equivalent given `assert data.target_epoch in (get_previous_epoch(state), get_current_epoch(state))`. I prefer `if` because it's more symmetric :)

Reply:
This is not strictly equivalent. At the genesis epoch, `get_previous_epoch == get_current_epoch` and so you would add the `pending_attestation` to both `previous_epoch_attestations` and `current_epoch_attestations`

Reply:
Oooh right I forgot `get_previous_epoch == get_current_epoch` case! And... if so, this PR is a bugfix, right?

Reply:
No, the previous if/else handled it fine

Reply:
@djrtwo Ah sorry I confused myself. (😵)

Do you mean we should only update `current_epoch_attestations` at the genesis epoch?
```python
if data.target_epoch == get_current_epoch(state):
    ....
elif if data.target_epoch == get_previous_epoch(state):
    ....
```

Reply:
Is it okay to introduce `raise`?
```python
if data.target_epoch == get_current_epoch(state):
    ....
elif data.target_epoch == get_previous_epoch(state):
    ....
else:
    raise(Exception("Invalid target_epoch"))
```

Reply:
Fixed genesis bug :)